agent/php_environment.c

```diff
@@ -511,6 +511,124 @@ static void nr_php_get_environment_variables(TSRMLS_D) {
                    __func__, NR_PHP_PROCESS_GLOBALS(env_labels));
 }
 
+#define DOCKER_ID_V2_STRLEN (64)
+#define MAX_LINE_COUNT (1000)  // Upper bound for number of lines to read
+#define MAX_INNER_LOOP (10)    // Upper bound for number of token loops
+/*
+ * Purpose:
+ *    Extract the 64-byte hexadecimal Docker cgroup ID from
+ *     /proc/self/mountinfo
+ */
+char* nr_php_parse_v2_docker_id(const char* cgroup_fname) {
+  char* line_ptr = NULL;
+  char* token = NULL;
+  char* retval = NULL;
+  char* saveptr = NULL;
+  bool found = false;
+  int line_count = 0;
+  int token_loop_count = 0;
+  FILE* fd = NULL;
+  size_t len = 0;
+  nr_regex_t* regex = NULL;
+
+  if (NULL == cgroup_fname) {
+    return NULL;
+  }
+
+  // check if file exists
+  if (SUCCESS != access(cgroup_fname, F_OK)) {
+    nrl_verbosedebug(NRL_AGENT, "File not found: %s", cgroup_fname);
+    return NULL;
+  }
+
+  // open file
+  fd = fopen(cgroup_fname, "r");
+  if (NULL == fd) {
+    nrl_warning(NRL_AGENT, "Failed to open %s", cgroup_fname);
+    return NULL;
+  }
+
+  // compile regex to verify hexadecimal chars
+  regex = nr_regex_create("^[a-fA-F0-9]+$", 0, 0);
+
+  if (NULL == regex) {
+    nrl_error(NRL_AGENT, "Error: regex creation failed");
+    return NULL;
+  }
+
+  // clang-format off
+  /* 
+   * Example /proc/self/mountinfo file structure:
+   *   ...
+   *   795 787 254:1 /docker/containers/ec807d5258c06c355c07e2acb700f9029d820afe5836d6a7e19764773dc790f5/resolv.conf /etc/resolv.conf rw,relatime - ext4 /dev/vda1 rw
+   *   796 787 254:1 /docker/containers/ec807d5258c06c355c07e2acb700f9029d820afe5836d6a7e19764773dc790f5/hostname /etc/hostname rw,relatime - ext4 /dev/vda1 rw
+   *   797 787 254:1 /docker/containers/ec807d5258c06c355c07e2acb700f9029d820afe5836d6a7e19764773dc790f5/hosts /etc/hosts rw,relatime - ext4 /dev/vda1 rw
+   *   ...
+   */
+
+  /*
+   * File parsing logic:
+   *  1. scan file line-by-line
+   *  2. split each line into '/' delimited tokens
+   *  3. search for "docker" followed by "container" string tokens
+   *  4. extract the 64 byte field following "/docker/containers/"
+   *  5. Verify the extracted ID is both:
+   *    a. 64 bytes long (not including null terminator)
+   *    b. comprised of only hexadecimal characters
+   *  6. Assign the extracted & verified ID to the retval
+   *    a. Example ID: ec807d5258c06c355c07e2acb700f9029d820afe5836d6a7e19764773dc790f5
+   *  7. Set found = true and exit the loops
+   */
+  // clang-format on
+
+  while (FAILURE != getline(&line_ptr, &len, fd) && !found
+         && line_count++ < MAX_LINE_COUNT) {
+    token = strtok_r(line_ptr, "/", &saveptr);
```

Comment:
Is it more efficient to tokenize the line by `/` vs making the regex something like:
`regex = nr_regex_create("^/docker/containers/([a-fA-F0-9]{DOCKER_ID_V2_STRLEN})/", 0, 0);`

Then using [nr_regex_match_capture and nr_regex_substrings_get](https://github.com/newrelic/newrelic-php-agent/blob/main/axiom/tests/test_regex.c#L145) to get the ID?

It could save the tokenizing step and the strlen comparison as well.

Reply:
`nr_regex_match_capture` requires a `const char* str` as it's second argument- what would that be if we don't know the actual string value of the id in advance? Looking at the examples, it seems they all require an explicit string containing the value to be passed, is this not the case?

Reply:
You can pass the modified regex and line_ptr in and match_capture will pull a valid docker id out.

[Wordpress](https://github.com/newrelic/newrelic-php-agent/blob/main/agent/fw_wordpress.c#L436) uses this as well.

Reply:
Ok I think I understand- my previous understanding of this function family was incorrect. I'll investigate this implementation and see if I can coerce it into to working for me

Reply:
Valgrind really despises trying to pass off the `line_ptr` as a `const char*` and is complaining about "Conditional jump depends on uninitialized values" in `nr_regex_match_capture`

Reply:
How are you declaring/passing in line_ptr?

The wordpress example just declares a `char *` [here](https://github.com/newrelic/newrelic-php-agent/blob/main/agent/fw_wordpress.c#L422) then passes it in [here](https://github.com/newrelic/newrelic-php-agent/blob/main/agent/fw_wordpress.c#L437)

Maybe it would prefer you to [strdup](https://github.com/newrelic/newrelic-php-agent/blob/main/agent/fw_wordpress.c#L437) it first?

Reply:
Tried passing `line_ptr` normally, tried casting it to `const char*`, tried `strdup`/`strndup`ing it, no luck. 
I think the critical difference between this case and the one in wordpress is that the `char *` is still based on the `const` zval `tag` argument.

Reply:
Commenting for posterity: The issue was with trying to use the `len` value set by the `getline` function. Using an explicit `nr_strlen` on the `line_ptr` instead solved the valgrind issue.

Reply:
Fixed in 65bd871c117f5dddde7813c8e3f1aec9c04558e7